state.go

```diff
@@ -356,24 +356,52 @@ func (i *InitState) Handle(ctx *StateContext, c Controller) (State, bool) {
 
 	log.Infof("handling loaded state: %s", sd.Name)
 
-	// chack last known state
-	switch sd.Name {
-	// update process was finished; check what is the status of update
-	case MenderStateReboot:
+	if !committedPartition(c) {
+		// only valid entrypoint into the uncommitted partition is reboot_leave
+		if sd.Name != MenderStateReboot {
+			// entered the uncommitted partition without finishing the whole update-path
+			// on the committed partition, therefore reboot back into the committed partition
+			// and error
+			if err := c.Reboot(); err != nil {
+				return NewUpdateErrorState(NewFatalError(errors.Errorf("failed to reboot device: %v", err)), sd.UpdateInfo), false
+			}
+			// should never happen
+			return doneState, false
+		}
 		return NewAfterRebootState(sd.UpdateInfo), false
+	}
+
+	// check last known state
+	switch sd.Name {
 
 	case MenderStateRollbackReboot:
 		return NewAfterRollbackRebootState(sd.UpdateInfo), false
 
-	// this should not happen
+	// Rerun commit-leave
+	case MenderStateUpdateCommit:
+		c.SetNextState(NewUpdateCommitState(sd.UpdateInfo))
+		return idleState, false
+
+	// invalid entrypoint into the state-machine. Error out.
 	default:
-		log.Errorf("got invalid state: %v", sd.Name)
+		log.Errorf("got invalid entrypoint into the state machine: state: %v", sd.Name)
 		me := NewFatalError(errors.Errorf("got invalid state stored: %v", sd.Name))
 
 		return NewUpdateErrorState(me, sd.UpdateInfo), false
 	}
 }
 
+func committedPartition(c Controller) bool {
+
+	ua, err := c.HasUpgrade()
+	log.Infof("in committedPartition: ua: %t, err: %v", ua, err)
+	if err != nil {
```

Comment:
Maybe we can print some error message here as this is pretty important and will impact state transition.